package/network/utils/umbim/files/lib/netifd/proto/mbim.sh

```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
 		proto_notify_error "$interface" NO_REGISTRATION
 		return 1
 	}
 	tid=$((tid + 1))
 
 	echo "mbim[$$]" "Attach to network"
-   	umbim $DBG -n -t $tid -d $device attach || {
+	umbim $DBG -n -t $tid -d $device attach || {
 		echo "mbim[$$]" "Failed to attach to network"
 		proto_notify_error "$interface" ATTACH_FAILED
 		return 1
 	}
 	tid=$((tid + 1))
- 
+
 	echo "mbim[$$]" "Connect to network"
-	while ! umbim $DBG -n -t $tid -d $device connect "$apn" "$auth" "$username" "$password"; do
+	while ! umbim $DBG -n -t $tid -d $device connect "$pdptype:$apn" "$auth" "$username" "$password"; do
 		tid=$((tid + 1))
 		sleep 1;
 	done
 	tid=$((tid + 1))
 
+	echo "mbim[$$]" "Connected, obtain IP address and configure interface"
+	config="/tmp/mbim.$$.config"
+	umbim $DBG -n -t $tid -d $device config > "$config" || {
+		echo "mbim[$$]" "Failed to obtain IP address"
+		proto_notify_error "$interface" CONFIG_FAILED
+		rm -f "$config"
+		return 1
+	}
+	cat "$config"
+	tid=$((tid + 1))
+
 	uci_set_state network $interface tid "$tid"
 
-	echo "mbim[$$]" "Connected, starting DHCP"
+	local ip_4 ip_6
+	ip_4=$(awk '$1=="ipv4address:" {print $2}' "$config")
+	ip_6=$(awk '$1=="ipv6address:" {print $2}' "$config")
+	[ "$ip_4" ] || [ "$ip_6" ] || {
+		echo "mbim[$$]" "Failed to obtain IP addresses"
+		proto_notify_error "$interface" CONFIG_FAILED
+		rm -f "$config"
+		return 1
+	}
+
 	proto_init_update "$ifname" 1
-	proto_send_update "$interface"
+	proto_set_keep 1
+	local ip mask gateway mtu dns1 dns2
+
+	[ "$ip_4" ] && {
+		echo "mbim[$$]" "Configure IPv4 on $ifname"
+		ip=${ip_4%%/*}
+		mask=${ip_4##*/}
+		gateway=$(awk '$1=="ipv4gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv4mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv4_address "$ip" "$mask"
+		[ "$defaultroute" = 0  ] || proto_add_ipv4_route 0.0.0.0 0 "$gateway" "$ip_4" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+	}
+	[ "$ip_6" ] && {
+		echo "mbim[$$]" "Configure IPv6 on $ifname"
+		ip=${ip_6%%/*}
+		mask=${ip_6##*/}
+		gateway=$(awk '$1=="ipv6gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv6mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv6_address "$ip" "$mask"
+		proto_add_ipv6_prefix "$ip_6"
+		[ "$defaultroute" = 0  ] || proto_add_ipv6_route "::" 0 "$gateway" "" "" "$ip_6" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+		# TODO: Don't know how to propagate IPv6 address to lan from here?
+	}
 
```

Comment:
This looks nice to me, but I am no expert in netifd or OpenWrt. 

I guess you would want to optionally plug the IPv6 prefix into odhcpd here too?

Reply:
the comment is wrong, everything is being propagated just fine, i actually can't seem to figure out why clients can't ping the outside world, the routes all look right to me, client configurations look perfect, i can't explain it